scvi/external/scar/_module.py

```diff
@@ -0,0 +1,460 @@
+            n_layers=n_layers,
+            n_hidden=n_hidden,
+            dropout_rate=0,
+            inject_covariates=inject_covariates,
+            use_batch_norm=use_batch_norm,
+            use_layer_norm=use_layer_norm,
+        )
+
+        # mean gamma
+        if scale_activation == "softmax":
+            px_scale_activation = nn.Softmax(dim=-1)
+        elif scale_activation == "softplus":
+            px_scale_activation = nn.Softplus()
+        elif scale_activation == "softplus_sp":
+            px_scale_activation = softplus(sparsity)
+        self.px_scale_decoder = nn.Sequential(
+            nn.Linear(n_hidden, n_output),
+            px_scale_activation,
+            hnormalization(),
+        )
+
+        # noise ratio
+        self.px_noise_decoder = nn.Sequential(
+            nn.Linear(n_hidden, 1),
+            tanh(),
+        )
+
+        # dispersion: here we only deal with gene-cell dispersion case
+        self.px_r_decoder = nn.Linear(n_hidden, n_output)
+
+        # dropout
+        self.px_dropout_decoder = nn.Linear(n_hidden, n_output)
+
+    def forward(
+        self,
+        dispersion: str,
+        z: torch.Tensor,
+        library: torch.Tensor,
+        *cat_list: int,
+    ):
+        """
+        The forward computation for a single sample.
+
+         #. Decodes the data from the latent space using the decoder network
+         #. Returns parameters for the ZINB distribution of expression
+         #. If ``dispersion != 'gene-cell'`` then value for that param will be ``None``
+
+        Parameters
+        ----------
+        dispersion
+            One of the following
+
+            * ``'gene'`` - dispersion parameter of NB is constant per gene across cells
+            * ``'gene-batch'`` - dispersion can differ between different batches
+            * ``'gene-label'`` - dispersion can differ between different labels
+            * ``'gene-cell'`` - dispersion can differ for every gene in every cell
+        z :
+            tensor with shape ``(n_input,)``
+        library_size
+            library size
+        cat_list
+            list of category membership(s) for this sample
+
+        Returns
+        -------
+        4-tuple of :py:class:`torch.Tensor`
+            parameters for the ZINB distribution of expression
+
+        """
+        # The decoder returns values for the parameters of the ZINB distribution
+        px = self.px_decoder(z, *cat_list)
+        px_scale = self.px_scale_decoder(px)
+        px_dropout = self.px_dropout_decoder(px)
+
+        # noise ratio
+        px_noise_ratio = self.px_noise_decoder(px)
+
+        # Clamp to high value: exp(12) ~ 160000 to avoid nans (computational stability)
+        px_rate = torch.exp(library) * px_scale  # torch.clamp( , max=12)
+        px_r = self.px_r_decoder(px) if dispersion == "gene-cell" else None
+
+        return px_scale, px_noise_ratio, px_r, px_rate, px_dropout
+
+
+class SCAR_VAE(VAE):
```

Comment:
Are the architectures exactly matched? If not, can you comment on this in the docstring of the model as well as in the user guide? At least this way people know it's not exactly the same but the outputs were checked to be similar on some examples.